provisioner/clusterpy.go

```diff
@@ -333,38 +334,73 @@ func (p *clusterpyProvisioner) Provision(ctx context.Context, logger *log.Entry,
 		azInfo:        azInfo,
 	}
 
+	karpenterProvisioner, err := NewKarpenterNodePoolProvisioner(
+		NodePoolTemplateRenderer{
+			awsAdapter:     awsAdapter,
+			config:         channelConfig,
+			cluster:        cluster,
+			bucketName:     bucketName,
+			logger:         logger,
+			encodeUserData: false,
+		}, p.execManager, p.tokenSource,
+	)
+	if err != nil {
+		return err
+	}
+
 	// group node pools based on their profile e.g. master
 	nodePoolGroups := groupNodePools(
 		logger,
 		cluster,
-		nodePoolProvisioner,
+		caNodePoolProvisioner,
+		karpenterProvisioner,
 	)
 
 	err = nodePoolGroups["masters"].provisionNodePoolGroup(ctx, values, updater, cluster, p.applyOnly)
 	if err != nil {
 		return err
 	}
 
-	if err = ctx.Err(); err != nil {
-		return err
+	//if err = ctx.Err(); err != nil {
+	//	return err
+	//}
+
+	if karpenterProvisioner.isKarpenterEnabled() {
+		err = p.apply(ctx, logger, cluster, deletions, manifests)
```

Comment:
@linki @myaser I remember we discussed this back in 2022 and we thought about looking at previous kubernetes-on-aws changes where the order (node rotation, apply manifests) mattered, to see if changing this order would be a problem.

Did we do this?

Reply:
Not really
We would need a list of typical updates and then check if an order change could cause a problem
I don't think we have such a list

do we audit log zregistry commands? would be the place where we build this list, otherwise it will be only guessing

Reply:
For these more complex updates we often the naming pattern of **Change X [1/2]** or similar:

E.g. like these:

![image](https://user-images.githubusercontent.com/128566/230391746-31c3049c-4513-45e3-8d4c-840a4aa91af3.png)

My proposal would be to look at such PRs and understand if they would work or if the order could be constructed differently such that they would. It want find _all_ cases but can be a good check to think about the potential consequences.

(It seems not easy to search for this pattern in github search though 😥)

Reply:
using github CLI to list all PRs
```bash
gh pr list --state all --limit 5777
```
and filtering the output using regex, I compiled this list of 61 occurance
```
5663	Enable 2k nodes by default stage [2/2]	enable-2k-nodes-stage2	MERGED	2023-01-09 17:02:37 +0000 UTC
5643	Enable 2k nodes by default stage1 [1/2]	enable-2k-nodes-stage1	MERGED	2023-01-03 11:12:47 +0000 UTC
5150	Consolidate e2e application [1/2]	cleanup-consolidation-1	MERGED	2022-05-30 16:27:59 +0000 UTC
4990	Consolidate DaemonSet applications/components [3/3]	consolidate-ds-apps-3	MERGED	2022-02-14 15:03:52 +0000 UTC
4975	Consolidate DaemonSet applications/components [2/3]	consolidate-ds-apps-2	MERGED	2022-02-08 12:43:24 +0000 UTC
4945	Consolidate prometheus into kubernetes application [3/3]	prometheus-consolidation-3	MERGED	2022-01-28 09:33:55 +0000 UTC
4939	Set custom cluster tag on egress stacks [2/2]	egress-controller-cluster-id-tag	MERGED	2022-01-26 14:43:29 +0000 UTC
4934	Consolidate prometheus into kubernetes application [2/3]	prometheus-consolidation-2	MERGED	2022-01-25 16:05:30 +0000 UTC
4910	Set custom cluster tag on egress stacks [1/2]	egress-cluster-tag	MERGED	2022-01-19 14:33:09 +0000 UTC
4900	Consolidate prometheus into kubernetes application [1/3]	prometheus-consolidation	MERGED	2022-01-14 17:38:47 +0000 UTC
4898	Consolidate DaemonSet applications/components [1/3]	consolidate-ds-apps	MERGED	2022-01-14 16:43:38 +0000 UTC
4704	Skipper/routesrv2 [2/2]	skipper/routesrv2	MERGED	2021-10-06 16:33:14 +0000 UTC
4702	Skipper/routesrv [1/2]	skipper/routesrv	MERGED	2021-10-05 18:10:07 +0000 UTC
4393	Remove spotio node pool from e2e [1/2]	rm-spotio-0	MERGED	2021-06-23 11:33:16 +0000 UTC
4392	Remove spotio support [2/2]	rm-spotio	MERGED	2021-06-23 10:10:07 +0000 UTC
4327	Drop worker_shared_secret from the e2e script [2/2]	drop-shared-secret-2	MERGED	2021-05-31 15:51:52 +0000 UTC
4326	Drop node_auth and worker_shared_secret from the manifests [1/2]	drop-shared-secret	MERGED	2021-05-31 15:50:21 +0000 UTC
4284	credentials-provider: fully migrate to restricted tokens [2/2]	credprov-no-old-tokens	MERGED	2021-05-05 12:48:05 +0000 UTC
4260	node_auth: enabled->exclusive [2/2]	node-auth-exclusive	MERGED	2021-04-28 10:49:23 +0000 UTC
4259	node_auth: supported->enabled [1/2]	node-auth-enabled	MERGED	2021-04-28 10:47:19 +0000 UTC
4243	[2/3] Reduce skipper metrics dimension	skipper-new-counter-metric	MERGED	2021-04-19 17:20:39 +0000 UTC
4235	[1/3] Update skipper and use new counter metric	skipper-new-counter-metric	MERGED	2021-04-19 05:47:48 +0000 UTC
4201	Master SG: drop the ingress rules that were only used for metrics [2/2]	close-unnecessary-ports	MERGED	2021-03-29 16:11:57 +0000 UTC
4199	Cleanup the metrics for the apiserver components [1/2]	apiserver-container-metrics	MERGED	2021-03-29 15:59:50 +0000 UTC
4112	E2E: generate a per-run WORKER_SHARED_SECRET [3/3]	e2e-auth-cleanup	MERGED	2021-03-04 15:10:12 +0000 UTC
4109	E2E: support OAuth tokens as well [1/3]	e2e-auth-pre	MERGED	2021-03-04 11:59:17 +0000 UTC
4108	E2E: use an OAuth token instead of a static secret [2/3]	e2e-auth	MERGED	2021-03-04 11:51:00 +0000 UTC
4092	Support scraping Prometheus metrics via pod entities [2/2]	prometheus-pod-entity-enable	MERGED	2021-02-25 15:02:07 +0000 UTC
4086	Support scraping Prometheus metrics via pod entities [1/2]	prometheus-pod-entities	MERGED	2021-02-25 08:16:52 +0000 UTC
4013	Switch back to legacy OIDC path for migration [2/2]	add-legacy-oidc-keys-path-part-2	MERGED	2021-02-03 15:48:06 +0000 UTC
4011	Still support old OIDC keys path [1/2]	add-legacy-oidc-keys-path	MERGED	2021-02-03 14:43:10 +0000 UTC
3947	Add kube-node-ready-controller [2/2] (test)	enable-kube-node-ready-controller-test	MERGED	2021-01-19 09:34:06 +0000 UTC
3907	Add kube-node-ready-controller [1/2]	kube-node-ready-controller	MERGED	2021-01-08 16:49:01 +0000 UTC
3886	Improve the e2e steps [2/2]	better-certs-2	MERGED	2021-01-06 15:56:46 +0000 UTC
3885	Improve the e2e steps [1/2]	better-certs	MERGED	2021-01-06 12:43:39 +0000 UTC
3871	Drop admission controller proxy [2/2]	drop-proxy-2	MERGED	2021-01-04 10:46:52 +0000 UTC
3870	Drop admission controller proxy [1/2]	drop-proxy-1	MERGED	2021-01-04 10:31:51 +0000 UTC
3700	Support permanent scale-up for CA v1.18 [1/2]	cluster_autoscaler_1_18_2	CLOSED	2020-10-30 09:56:45 +0000 UTC
3396	Remove all traces of API server ELBv1 (part 2/2)	remove-elb-entirely	MERGED	2020-07-07 15:13:44 +0000 UTC
3393	Remove all traces of API server ELBv1 (part 1.9/2)	remove-elb-entirely	MERGED	2020-07-07 12:35:34 +0000 UTC
3373	Remove all traces of API server ELBv1 (part 1/2)	remove-elbv1	MERGED	2020-07-01 10:08:32 +0000 UTC
2891	[1/2] CDP controller role: allow the use of v2 tokens	cdp-controller-v2	MERGED	2020-01-23 12:16:29 +0000 UTC
2865	RBAC: rollout [2/2]	revert-2864-revert-2862-revert-2861-revert-2456-webhook-deny-only	MERGED	2020-01-14 17:51:23 +0000 UTC
2864	Revert "Revert "Revert "RBAC: rollout [2/2]"""	revert-2862-revert-2861-revert-2456-webhook-deny-only	MERGED	2020-01-14 15:22:25 +0000 UTC
2862	Revert "Revert "RBAC: rollout [2/2]""	revert-2861-revert-2456-webhook-deny-only	MERGED	2020-01-14 12:59:01 +0000 UTC
2861	Revert "RBAC: rollout [2/2]"	revert-2456-webhook-deny-only	MERGED	2020-01-14 11:14:16 +0000 UTC
2844	Update audittrail-adapter to work with v1 audit events instead of v1beta1 [1/2]	audittrail-adapter-new-api	MERGED	2020-01-08 13:14:09 +0000 UTC
2749	Limit nodes considered for ingress load balancer based on taints [1/2]	ingress-controller-targets	MERGED	2019-12-04 16:11:29 +0000 UTC
2520	Revert "logging-agent: support new node configuration [1/2]"	revert-logging-change	MERGED	2019-09-16 13:14:07 +0000 UTC
2515	logging-agent: support new node configuration [1/2]	migrate-logging-agent	MERGED	2019-09-13 16:02:45 +0000 UTC
2510	RBAC: rollout [1/2]	update-rbac-roles	MERGED	2019-09-12 11:11:18 +0000 UTC
2456	RBAC: rollout [2/2]	webhook-deny-only	MERGED	2019-08-27 15:41:35 +0000 UTC
2202	RBAC zalando iam 2 [2/2]	rbac-zalando-iam-2	MERGED	2019-06-06 09:57:18 +0000 UTC
2200	Add RBAC for Zalando IAM users [1/2]	rbac-zalando-iam	MERGED	2019-06-05 13:58:41 +0000 UTC
2170	Switch kubelets to authenticated mode [3/3]	kubelet-auth-part-3	MERGED	2019-05-27 09:34:57 +0000 UTC
2167	Kubelet: use authenticated API by default [2/3]	kubelet-auth-part-2	MERGED	2019-05-24 08:25:57 +0000 UTC
2153	Kubelet: switch to config instead of arguments, use the new certificate [1/3]	kubelet-auth-part-1	MERGED	2019-05-22 10:17:56 +0000 UTC
1183	dnsmasq-node to 1.9 (production 1/2)	dnsmasq-to-1.9	MERGED	2018-06-19 12:02:27 +0000 UTC
1181	Flannel to 1.9 (production 2/2)	flannel-to-1.9	MERGED	2018-06-19 10:20:56 +0000 UTC
1121	Remove obsolete stack parameters (Legacy pools step 4/4)	remove-static-pools2	MERGED	2018-06-05 10:06:39 +0000 UTC
1091	Remove obsolete static node pools (Legacy pools step 2/4)	remove-static-pools	MERGED	2018-05-30 13:10:11 +0000 UTC
```